src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ManagedTelemetryService.cs

```diff
@@ -3,81 +3,139 @@
 using System.Security.Cryptography;
 using System.Text;
 
-namespace Microsoft.VisualStudio.Telemetry
+namespace Microsoft.VisualStudio.Telemetry;
+
+[Export(typeof(ITelemetryService))]
+internal class ManagedTelemetryService : ITelemetryService
 {
-    [Export(typeof(ITelemetryService))]
-    internal class ManagedTelemetryService : ITelemetryService
+#if DEBUG
+    private const string EventNamePrefix = "vs/projectsystem/managed/";
+    private const string PropertyNamePrefix = "vs.projectsystem.managed.";
+#endif
+
+    public void PostEvent(string eventName)
     {
-        public void PostEvent(string eventName)
-        {
-            Requires.NotNullOrEmpty(eventName);
+        Requires.NotNullOrEmpty(eventName);
 
-            PostTelemetryEvent(new TelemetryEvent(eventName));
-        }
+        PostTelemetryEvent(new TelemetryEvent(eventName));
+    }
+
+    public void PostProperty(string eventName, string propertyName, object propertyValue)
+    {
+        Requires.NotNullOrEmpty(eventName);
+        Requires.NotNullOrEmpty(propertyName);
+        Requires.NotNull(propertyValue);
+
+        TelemetryEvent telemetryEvent = new(eventName);
+        telemetryEvent.Properties.Add(propertyName, propertyValue);
+
+        PostTelemetryEvent(telemetryEvent);
+    }
+
+    public void PostProperties(string eventName, IEnumerable<(string propertyName, object propertyValue)> properties)
+    {
+        Requires.NotNullOrEmpty(eventName);
+        Requires.NotNullOrEmpty(properties);
 
-        public void PostProperty(string eventName, string propertyName, object propertyValue)
+        TelemetryEvent telemetryEvent = new(eventName);
+        AddPropertiesToEvent(properties, telemetryEvent);
+
+        PostTelemetryEvent(telemetryEvent);
+    }
+
+    private static void AddPropertiesToEvent(IEnumerable<(string propertyName, object propertyValue)> properties, TelemetryEvent telemetryEvent)
+    {
+        foreach ((string propertyName, object propertyValue) in properties)
         {
-            Requires.NotNullOrEmpty(eventName);
-            Requires.NotNullOrEmpty(propertyName);
-            Requires.NotNull(propertyValue);
+            if (propertyValue is ComplexPropertyValue complexProperty)
+            {
+                telemetryEvent.Properties.Add(propertyName, new TelemetryComplexProperty(complexProperty.Data));
```

Comment:
Is there a reason PostProperty (single edition) doesn't check for complex property value?